File: python/08_067.py
class BoatNode:

    def __init__(self, id, maintenanceNeeded, next=None):
        self.id = id
        self.maintenanceNeeded = maintenanceNeeded
        self.next = next

def reorderFleet(head):
    if not head or not head.next:
        return head
    
    # Create dummy heads for our two chains
    maintenance_dummy = BoatNode(0, False)
    no_maintenance_dummy = BoatNode(0, False)
    
    # Pointers to track the end of each chain
    maintenance_tail = maintenance_dummy
    no_maintenance_tail = no_maintenance_dummy
    
    # Traverse the original list
    current = head
    while current:
        next_boat = current.next
        current.next = None  # Detach the node
        
        if current.maintenanceNeeded:
            maintenance_tail.next = current
            maintenance_tail = current
        else:
            no_maintenance_tail.next = current
            no_maintenance_tail = current
        
        current = next_boat
    
    # Connect the two chains
    maintenance_tail.next = no_maintenance_dummy.next
    
    # Return the head of the merged chain
    return maintenance_dummy.next

File: data structures and algorithms/cs/08_067.cs
using System;

class BoatNode
{
   public int Id { get; }
   public bool MaintenanceNeeded { get; }
   public BoatNode Next { get; set; }

   public BoatNode(int id, bool maintenanceNeeded)
   {
      Id = id;
      MaintenanceNeeded = maintenanceNeeded;
      Next = null;
   }
}

public class Solution
{
   public BoatNode ReorderFleet(BoatNode head)
   {
      if (head == null || head.Next == null)
      {
         return head;
      }

      var maintenanceDummy = new BoatNode(0, false);
      var noMaintenanceDummy = new BoatNode(0, false);

      var maintenanceTail = maintenanceDummy;
      var noMaintenanceTail = noMaintenanceDummy;

      BoatNode current = head;
      while (current != null)
      {
         BoatNode nextBoat = current.Next;
         current.Next = null;  // Detach the node

         if (current.MaintenanceNeeded)
         {
            maintenanceTail.Next = current;
            maintenanceTail = current;
         }
         else
         {
            noMaintenanceTail.Next = current;
            noMaintenanceTail = current;
         }

         current = nextBoat;
      }

      maintenanceTail.Next = noMaintenanceDummy.Next;

      return maintenanceDummy.Next;
   }
}
